testdata/project-v3-config/Makefile

```diff
@@ -68,14 +68,40 @@ build: generate fmt vet ## Build manager binary.
 run: manifests generate fmt vet ## Run a controller from your host.
 	go run ./main.go
 
+# The build is supported with BuildKit enabled
+# If you disable this option then the manager image will be built using linux/amd64 by default
+# Alternatively, you can customize the Dockerfile by removing the --platform flag
+# However, in this case you will be unable to use the docker-buildx target to build with cross-compilation 
+# the manager to support multiple architectures
+export DOCKER_BUILDKIT=1
+
+# To get the system architecture to allow the build work if/when buildKit is disabled
+LOCALPLATFORM ?= linux/$(shell uname -m | sed 's/x86_64/amd64/' | sed 's/aarch64/arm64/')
```

Comment:
Same consideration for pulling these from go

Reply:
@jaypoulz @camilamacedo86 
@aleskandro and I had an offline conversation about this. 
Our conclusion is to remove `--platform` from the Dockerfile, eliminating the need for `LOCALPLATFORM` 
ie:
```
FROM golang:1.19 as builder
```
Basically `docker build .` should always build the architecture for the machine that is runs this command w/o the need to pass a variable, use make, or needing buildkit to be enabled. 

The special case here is cross builds.
We suggest adding logic in the `docker-buildx` section in the Makefile to insert `--platform` on the first `FROM ... as builder` .. so something like:

```
docker-buildx: test ## Build and push docker image for the manager for cross-platform support
        # copy existing Dockerfile and Insert --platform=${BUILDPLATFORM} into Dockerfile.cross, and preserve the original Dockerfile
        sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/'  Dockerfile > Dockerfile.cross
        
        - docker buildx create --name project-v4-builder
        docker buildx use project-v4-builder
        - docker buildx build --push --platform=$(PLATFORMS) --tag ${IMG} . -f Dockerfile.cross
        docker buildx rm project-v4-builder
        #remove Dockerfile.cross (or add it to gitignore)
        rm Dockerfile.cross 
 ```

Reply:
Not sure if applicable: there seems like a "cross-compilation" strategy that makes it possible to stick on `BUILDPLATFORM` for the build stage, and apply to multiple platforms in runtime stage.

References:
- https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide
- https://gist.github.com/tonistiigi/bd31caaf1ea4ec8e00aec1f47f76fb66

So that by dropping `--platform`, the `FROM` sentence can be:
```
FROM --platform=$BUILDPLATFORM golang:1.19 AS build
```

Reply:
Hi @Kavinjsir, 

The problem with  `--platform=$BUILDPLATFORM` and the reason we added `FROM --platform=${BUILDPLATFORM:-LOCALPLATFORM} golang:1.19 as builder` is because if/when the buildKit is disabled the BUILDPLATFORM is nil/empty. So, if the BUILDPLATFORM is empty we add the LOCALPLATFORM value that is informed via arg, 

HI @jeffdyoung,

For the project users, we will have by default the requirement to use build it. Which is enabled by default in the Dockerfile.  However, to allow them still able to run `make dockerbuild` when it is enabled we have the implementation:

Dockerfiler
> FROM --platform=${BUILDPLATFORM:-LOCALPLATFORM} golang:1.19 as builder

Makefile
```
# To get the system architecture to allow the build work if/when buildKit is disabled
LOCALPLATFORM ?= linux/$(shell uname -m | sed 's/x86_64/amd64/' | sed 's/aarch64/arm64/')
# If you wish built the manager image targeting other platforms you can use the --platform flag.
# (i.e. docker build --platform linux/arm64 ). However, you must enable docker buildKit for it.
# More info: https://docs.docker.com/develop/develop-images/build_enhancements/
.PHONY: docker-build
docker-build: test ## Build docker image with the manager.
	docker build -t ${IMG} .
	docker build --build-arg=LOCALPLATFORM=$(LOCALPLATFORM) -t ${IMG} .
```

The above implementation mainly is only with the purpose of not letting the build not work in CI/e2e tests. The official recommendation will be to use buildKit. **However, users are always able to change their code as please them as well.** 

Regards your suggestions: 

```sh
docker-buildx: test ## Build and push docker image for the manager for cross-platform support
        # copy existing Dockerfile and Insert --platform=${BUILDPLATFORM} into Dockerfile.cross, and preserve the original Dockerfile
        sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/'  Dockerfile > Dockerfile.cross
        
        - docker buildx create --name project-v4-builder
        docker buildx use project-v4-builder
        - docker buildx build --push --platform=$(PLATFORMS) --tag ${IMG} . -f Dockerfile.cross
        docker buildx rm project-v4-builder
        #remove Dockerfile.cross (or add it to gitignore)
        rm Dockerfile.cross 
```

I do not like the proposed solution to create another Dockerfile and that reduces a lot the maintainability and readability (ihmo). I would vote in keep the current implementation.

Reply:
Hello Camila, we were trying multiple "cases" with Jeff and came out with that to be the most efficient solution to consider all the requirements we defined together.
With efficient, I mean the one that gives the best value at the minimum cost (the slightly reduced elegance).

Even though the recommendation will be BUILDKIT for the devs environment, it is not 100% guaranteed in most of the kubernetes-based CI environments at the moment. It is not even said that the container engine is Docker or that the user is allowed to enable BUILDKIT in the nodes. We should consider Mac and Linux envs, dev envs, and build/CI envs.

The requirements we considered are:
- Being able to natively build an operator image targeting the same ISA that is used by the builder node (#2850);
- The build has to work regardless of the build engine settings, i.e., compatible with non-BuildKit envs: only MacOS enables it by default; 
- The build has to work regardless of the use of the Makefile targets (https://github.com/kubernetes-sigs/kubebuilder/issues/2850): without a default value for BUILDPLATFORM, `docker build` would fail without buildkit, and you had to change a test case for this reason because the env in Prow doesn't enable buildkit;
- The build with buildx has to continue being based on the cross-compilation strategy you used.

The way Jeff and I see to guarantee all these four points is to get rid of the `--platform=$BUILDPLATFORM` where it is not needed, i.e., when you're building a single manifest with `make docker-build` or the standalone `docker build .`.

If the sed command is not so readable, we could make it simpler with `awk '!x{x=sub("^FROM","FROM --platform=$${BUILDPLATFORM}")}1' Dockerfile > Dockerfile.cross`. (does it look simpler? Note that it's only tested on Linux)

If the issue is about temporarily creating a new Dockerfile, I'd say it's not as intrusive as creating a new builder instance every time you run `make buildx`. 

> The implementation mainly is only with the purpose of not letting the build not work in CI/e2e tests. 

Currently, we are changing a test to make it work. The solution above should work for the current cases, natively in any arch, and would add the `buildx` capabilities to the framework.

Reply:
HI @aleskandro,

Following are some comments inline:

The tool is to help out users create their projects and skill and speed up.
One of the most important points is to ensure the maintainability and readability of the code produced by default. Not efficient at all. 

> Even though the recommendation will be BUILDKIT for the devs environment, it is not 100% guaranteed in most of the kubernetes-based CI environments at the moment. 

The current solution works with or without buildKit enabled. So that will work in any CI as it has been working.
So, the suggestion made seems more like a matter of preference and the code does not look for what is more important for us. 

Therefore, it is important to highlight that for the Release process usually, authors will adopt Goreleaer ( which works with docker buildx ) However, they are able to change and customize the default scaffold as please them. What most matters for us are they be able to follow the quick start and test it locally initially and have good, easy examples of how to achieve the goal.

Sorry, but I still have not seen any raised that justifies we do the change as suggested. It seems more a matter of preference.

Reply:
done as suggested thank you.